src/Client/GetHedgedConnections.cpp

```diff
@@ -0,0 +1,491 @@
+#if defined(OS_LINUX)
+
+#include <Client/GetHedgedConnections.h>
+#include <Common/typeid_cast.h>
+
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+    extern const int LOGICAL_ERROR;
+    extern const int ALL_CONNECTION_TRIES_FAILED;
+    extern const int ALL_REPLICAS_ARE_STALE;
+}
+
+GetHedgedConnections::GetHedgedConnections(
+    const ConnectionPoolWithFailoverPtr & pool_,
+    const Settings * settings_,
+    const ConnectionTimeouts & timeouts_,
+    std::shared_ptr<QualifiedTableName> table_to_check_)
+    : pool(pool_), settings(settings_), timeouts(timeouts_), table_to_check(table_to_check_), log(&Poco::Logger::get("GetHedgedConnections"))
+{
+    shuffled_pools = pool->getShuffledPools(settings);
+    for (size_t i = 0; i != shuffled_pools.size(); ++i)
+        try_get_connections.emplace_back(shuffled_pools[i].pool, &timeouts, settings, table_to_check.get(), log);
+
+    max_tries
+        = (settings ? size_t{settings->connections_with_failover_max_tries} : size_t{DBMS_CONNECTION_POOL_WITH_FAILOVER_DEFAULT_MAX_TRIES});
+
+    fallback_to_stale_replicas = settings ? settings->fallback_to_stale_replicas_for_distributed_queries : false;
+    entries_count = 0;
+    usable_count = 0;
+    failed_pools_count = 0;
+}
+
+GetHedgedConnections::~GetHedgedConnections()
+{
+    pool->updateSharedError(shuffled_pools);
+}
+
+std::vector<GetHedgedConnections::ReplicaStatePtr> GetHedgedConnections::getManyConnections(PoolMode pool_mode)
+{
+    size_t min_entries = (settings && settings->skip_unavailable_shards) ? 0 : 1;
+
+    size_t max_entries;
+    if (pool_mode == PoolMode::GET_ALL)
+    {
+        min_entries = shuffled_pools.size();
+        max_entries = shuffled_pools.size();
+    }
+    else if (pool_mode == PoolMode::GET_ONE)
+        max_entries = 1;
+    else if (pool_mode == PoolMode::GET_MANY)
+        max_entries = settings ? size_t(settings->max_parallel_replicas) : 1;
+    else
+        throw DB::Exception("Unknown pool allocation mode", DB::ErrorCodes::LOGICAL_ERROR);
+
+    std::vector<ReplicaStatePtr> replicas;
+    replicas.reserve(max_entries);
+    for (size_t i = 0; i != max_entries; ++i)
+    {
+        auto replica = getNextConnection(false);
```

Comment:
`non_blocking = false`
I think it is strange. Cause we are waiting all the connections happen.
So, it looks like hedged requests work only for first data packet, but all packets before are still processed in order, for each replica.

I would expect should get only connection here, and other stuff must be done async. Cause if we successfully got `hello` form replica, it has a good chance to be alive (though it may be slow, and hedging for first data packet is useful).
I don't even think we need to check here if connection is connected at all.